miteD/service/service.py

```diff
@@ -4,39 +4,63 @@
 from contextlib import suppress
 from json import loads, dumps, JSONDecodeError
 from nats.aio.client import Client as NATS
+from functools import wraps, partial
 
 from miteD.service.client import RemoteService
+from miteD.service.utils import get_members_if, format_version_str
 import miteD.service.response as response
 
 
-def parse_wrapped_endpoints(cls, *args, **kwargs):
-    wrapped = cls(*args, **kwargs)
+def is_rpc_method(method):
+    return getattr(method, '__rpc_method__', False)
+
+
+def is_notification_handler(method):
+    return getattr(method, '__notification_handler__', False)
+
+
+def parse_wrapped_endpoints(cls):
+    wrapped = cls()
     endpoints = {'*': {}}
-    for member in [getattr(wrapped, member_name) for member_name in dir(wrapped)]:
-        if callable(member):
-            if hasattr(member, '__rpc_name__') and hasattr(member, '__rpc_versions__'):
-                for version in [version.replace('.', '_') for version in member.__rpc_versions__]:
-                    members = endpoints.get(version, {})
-                    members[member.__rpc_name__] = member
-                    endpoints[version] = members
+    for member in get_members_if(is_rpc_method, wrapped):
+        for version in member.__rpc_versions__:
+            members = endpoints.setdefault(version, {})
+            members[member.__rpc_name__] = member
     return endpoints
 
 
-def rpc_service(name, versions, broker_urls=('nats://127.0.0.1:4222',)):
+def get_notification_handlers(cls):
+    wrapped = cls()
+    return get_members_if(is_notification_handler, wrapped)
+
+
+def rpc_service(
+        name,
+        versions,
+        broker_urls=('nats://127.0.0.1:4222',),
+        notification_topics=None,
+):
     def wrapper(cls):
+        class Notify():
+            pass
 
         class Service(object):
+            _layer = 'service'
```

Comment:
any particular reason these are class variables and not instance variables of this wrapped `Service` class?

Reply:
I think this code would work in either cases.